File: src/engine/EnigView.java
package engine;

import engine.OpenGL.EnigWindow;
import engine.OpenGL.ShaderProgram;

import static org.lwjgl.glfw.GLFW.*;

public abstract class EnigView {
	public EnigWindow window;

	public long loopStartTime;

	public float deltaTime;

	public long frameStartTime;

	/**
	 * creates a new main view
	 */
	public EnigView(EnigWindow swindow) {
		window = swindow;
	}

	public void runLoop() {
		loopStartTime = System.nanoTime();
		frameStartTime = loopStartTime;
		while (!glfwWindowShouldClose(window.id)) {
			long newTime = System.nanoTime();
			deltaTime = (float) (newTime - frameStartTime) / 1e9f;
			frameStartTime = newTime;
			if (loop()) {
				cleanUp();
				break;
			}
			window.update();
			cleanUp();
		}
		setDown();
	}

	/**
	 * cleans up at the end of a frame
	 */
	public void cleanUp() {
		ShaderProgram.disable();
		window.resetOffsets();
		glfwSwapBuffers(window.id);
		glfwPollEvents();
		EnigWindow.checkGLError();
		//window.update();
	}

	/**
	 * loop that gets called every frame
	 * @return if the view should close after this frame ends
	 */
	public abstract boolean loop();

	/**
	 * does any neccisary cleanup
	 */
	public void setDown() {};
}

File: src/main/kotlin/engine/EnigView.kt
package engine

import engine.opengl.EnigContext
import engine.opengl.EnigWindow
import engine.opengl.checkGLError

abstract class EnigView {

	open fun generateResources(window : EnigWindow) {
		EnigContext.enterScope()
	}

	/**
	 * cleans up at the end of a frame
	 */
	open fun cleanUpResources() {
		EnigContext.exitScope()
	}

	/**
	 * loop that gets called every frame
	 * @return if the view should close after this frame ends
	 */
	abstract fun loop(frameBirth : Long, dtime : Float) : Boolean

	open fun runIn(window : EnigWindow) {
		generateResources(window)
		window.runView { frameBirth, dtime ->
			loop(frameBirth, dtime)
		}
		cleanUpResources()
	}

	open fun runInGLSafe(window : EnigWindow) {
		try {
			runIn(window)
		} catch (t : Throwable) {
			t.printStackTrace()
			checkGLError()
		}
	}
}